File: test/epddl_domain_test.exs
defmodule EpddlDomainTest do
  use ExUnit.Case
  doctest EPDDL

  require Record

  test "Empty Domain Definition" do
    simpleDomain = "(define (domain foo))"
    domainDef = EPDDL.parse(simpleDomain)
    assert(Record.is_record(domainDef, :domain))
    assert(EPDDL.domain(domainDef, :id) != :undefined)
    assert(EPDDL.domain(domainDef, :id) == "foo")
  end

  test "Predicates" do
    domainStr = "(define (domain foo) (:predicates (at ?v) (location ?v ?l)))"
    domainDef = EPDDL.parse(domainStr)
    assert(EPDDL.domain(domainDef, :requirements) == [])
    assert(length(EPDDL.domain(domainDef, :predicates)) == 2)

    # Ensure predicate records are defined to our satisfaction
    [at, location] = EPDDL.domain(domainDef, :predicates)
    assert(EPDDL.predicate(at, :id) == "at")
    assert(EPDDL.predicate(at, :vars) == [{:var, "v", :undefined}])

    assert(EPDDL.predicate(location, :id) == "location")
    assert(EPDDL.predicate(location, :vars) == [{:var, "v", :undefined},
                                                {:var, "l", :undefined}])
  end

  test "Simple Action" do
    domainStr = "(define (domain foo) (:action bar :effect()))"
    domain = EPDDL.parse(domainStr)

    assert(length(EPDDL.domain(domain, :actions)) == 1)
    [bar] = EPDDL.domain(domain, :actions)
    assert(EPDDL.action(bar, :id) == "bar")
    assert(EPDDL.action(bar, :parameters) == [])
    assert(EPDDL.action(bar, :precondition) == true)
    assert(EPDDL.action(bar, :effect) == true)
  end

  test "Action Effects" do
    domainStr = "(define (domain foo) (:action bar :effect (and (at ?v) (at
  ?b) (not (loc ?c ?d)))))"
    domain = EPDDL.parse(domainStr)

    [bar] = EPDDL.domain(domain, :actions)

    assert(EPDDL.action(bar, :effect) != :undefined)

    {:and, [at1, at2, notExpr]} = EPDDL.action(bar, :effect)
    {:predicate, "at", ["v"]} = at1
    {:predicate, "at", ["b"]} = at2

    {:not, loc} = notExpr
    {:predicate, "loc", ["c", "d"]} = loc
  end
end

File: test/epddl_domain_test.erl
-module(epddl_domain_test).

-include("../include/epddl_types.hrl").

-include_lib("eunit/include/eunit.hrl").

empty_domain_test() ->
    SimpleDomain = "(define (domain foo))",
    DomainDef = epddl:parse(SimpleDomain),
    ?assert(is_record(DomainDef, domain)),
    ?assert(DomainDef#domain.id /= undefined),
    ?assert(DomainDef#domain.id == <<"foo">>).

simple_action_test() ->
    DomainStr = "(define (domain foo) (:action bar :effect ()))",
    Domain = epddl:parse(DomainStr),

    ?assert(length(Domain#domain.actions) == 1),
    [Bar] = Domain#domain.actions,
    ?assert(is_record(Bar, action)),

    ?assert(Bar#action.id == <<"bar">>),
    ?assert(Bar#action.parameters == []),
    ?assert(Bar#action.precondition == true),
    ?assert(Bar#action.effect == #effect{delta=true}).

action_effect_test() ->
    DomainStr = "(define (domain foo) (:action bar :effect (and (loc ?v) (loc ?b) (not (loc ?c ?d)))))",
    Domain = epddl:parse(DomainStr),

    [Bar] = Domain#domain.actions,
    ?assert(Bar#action.effect /= undefined),

    BarEffect = Bar#action.effect,
    ?assert(is_record(BarEffect, effect)),
    #effect{delta={'and', [Loc1, Loc2, NotExpr]}} = BarEffect,

    #effect{delta={predicate, <<"loc">>, [<<"v">>]}} = Loc1,
    #effect{delta={predicate, <<"loc">>, [<<"b">>]}} = Loc2,
    #effect{delta={'not', Loc}} = NotExpr,
    {predicate, <<"loc">>, [<<"c">>, <<"d">>]} = Loc.

action_parameters_test() ->
    DomainStr = "(define (domain foo) (:action bar :parameters (?a ?b ?c)))",
    Domain = epddl:parse(DomainStr),

    [Bar] = Domain#domain.actions,

    A = {var, <<"a">>, undefined},
    B = {var, <<"b">>, undefined},
    C = {var, <<"c">>, undefined},
    ?assert(Bar#action.parameters == [A, B, C]).

simple_action_precondition_test() ->
    DomainStr = "(define (domain foo) (:action bar :precondition (and (loc ?x ?y) (isCar ?y))))",
    Domain = epddl:parse(DomainStr),

    [Bar] = Domain#domain.actions,
    {'and', [Loc, IsCar]} = Bar#action.precondition,
    {predicate, <<"loc">>, [<<"x">>, <<"y">>]} = Loc,
    {predicate, <<"isCar">>, [<<"y">>]} = IsCar.
